src/auth/discord.rs

```diff
@@ -1 +1,404 @@
+    use itertools::Itertools;
+    let mut user_joined_events: Vec<String> = vec![];
+    let user_events = identities
+        .iter()
+        .filter_map(|i| {
+            if i.key.eq("discord-event") {
+                Some(i.value.as_str())
+            } else {
+                None
+            }
+        })
+        .collect_vec();
+
+    if user_events.is_empty() {
+        return Ok(vec![]);
+    }
+    for thread in user_events.iter() {
+        let event_member_list: Vec<String> = apis::event_members(thread).await?;
+        dbg!(&event_member_list);
+        if event_member_list.contains(&ud.user_id) {
+            user_joined_events.push(thread.to_string());
+        }
+        // TODO:
+        // Return Error if group administrator does not exist
+    }
+    //let user_joined_servers = apis::user_servers(ud.token.as_str()).await?;
+    //dbg!(&user_joined_servers);
+    // filter the user starred repos with input
+    Ok(user_joined_events
+        .into_iter()
+        .map(|event| fpm::user_group::UserIdentity {
+            key: "discord-event".to_string(),
+            value: event,
+        })
+        .collect())
+}
+pub mod apis {
+    #[derive(serde::Deserialize)]
+    pub struct DiscordAuthResp {
+        pub access_token: String,
+    }
+    // TODO: API to starred a repo on behalf of the user
+    // API Docs: https://discord.com/developers/docs/getting-started
+    //API EndPoints: https://github.com/GregTCLTK/Discord-Api-Endpoints/blob/master/Endpoints.md
+    // TODO: It can be stored in the request cookies
+    pub async fn user_details(token: &str) -> fpm::Result<(String, String)> {
+        // API Docs: https://discord.com/api/users/@me
+        // TODO: Handle paginated response
+        #[derive(serde::Deserialize)]
+        struct UserDetails {
+            username: String,
+            id: String,
+        }
+        let user_obj: UserDetails = fpm::auth::utils::get_api(
+            "https://discord.com/api/users/@me",
+            format!("{}{}", "Bearer ", token).as_str(),
+        )
+        .await?;
+
+        Ok((user_obj.username, user_obj.id))
+    }
+
+    //This API will only be used to get access token for discord
+    pub async fn discord_token(url: &str, redirect_url: &str, code: &str) -> fpm::Result<String> {
+        let client_id = match std::env::var("DISCORD_CLIENT_ID") {
+            Ok(id) => id,
+            Err(_e) => {
+                println!("WARN: DISCORD_CLIENT_ID not set");
+                // TODO: Need to change this approach later
+                "FPM_TEMP_DISCORD_CLIENT_ID".to_string()
+            }
+        };
+        let client_secret = match std::env::var("DISCORD_CLIENT_SECRET") {
+            Ok(secret) => secret,
+            Err(_e) => {
+                println!("WARN: DISCORD_CLIENT_SECRET not set");
+                // TODO: Need to change this approach later
+                "FPM_TEMP_DISCORD_CLIENT_SECRET".to_string()
+            }
```

Comment:
Return error if `DISCORD_CLIENT_ID` or `DISCORD_CLIENT_SECRET` are not present.

Reply:
done